runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -11584,6 +11589,101 @@ J9::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::CodeGenerator *c
+
+   startLoop->setStartInternalControlFlow();
+   endLoop->setEndInternalControlFlow();
+
+   // vectorized add in loop, 16 bytes per iteration
+   // use srcOffset for loop counter, add starting offset to length
+   // subtract 16 (xmm register size) to prevent reading/writing beyond beyond the end of the array
+   generateRegImmInstruction(SUB4RegImm4, node, length, 16, cg);
+   generateRegRegInstruction(ADD4RegReg, node, length, srcOffset, cg);
+
+   generateLabelInstruction(LABEL, node, startLoop, cg);
+   generateRegRegInstruction(CMP4RegReg, node, srcOffset, length, cg);
+   generateLabelInstruction(JGE4, node, endLoop, cg);
+
+   generateRegMemInstruction(MOVDQURegMem, node, xmmHighReg, generateX86MemoryReference(srcBuffer, srcOffset, 0, headerOffset, cg), cg);
+   // TODO; avoid this copy in the future by adding support for VPUNPCK instructions
+   generateRegRegInstruction(MOVDQURegReg, node, xmmLowReg, xmmHighReg, cg);
+   generateRegRegInstruction(PUNPCKHBWRegReg, node, xmmLowReg, zeroReg, cg);
+   generateMemRegInstruction(MOVDQUMemReg, node, generateX86MemoryReference(destBuffer, destOffset, 0, headerOffset + 16, cg), xmmLowReg, cg);
+
+   generateRegRegInstruction(PUNPCKLBWRegReg, node, xmmHighReg, zeroReg, cg);
+   generateMemRegInstruction(MOVDQUMemReg, node, generateX86MemoryReference(destBuffer, destOffset, 0, headerOffset, cg), xmmHighReg, cg);
+
+   // increase src offset by size of imm register
+   // increase dest offset by double, to account for the byte->char inflation
+   generateRegImmInstruction(ADD4RegImm4, node, srcOffset, 16, cg);
+   generateRegImmInstruction(ADD4RegImm4, node, destOffset, 32, cg);
+
+   //   LOOP BACK
+   generateLabelInstruction(JMP4, node, startLoop, cg);
+   generateLabelInstruction(LABEL, node, endLoop, cg);
+
+   // HANDLE REMAINDER (< 16 bytes)
+   TR::LabelSymbol *remStartLoop = generateLabelSymbol(cg);
+   TR::LabelSymbol *remEndLoop = generateLabelSymbol(cg);
+
+   remStartLoop->setStartInternalControlFlow();
+   remEndLoop->setEndInternalControlFlow();
+
+   generateRegImmInstruction(ADD4RegImm4, node, length, 16, cg);
+   generateRegRegInstruction(XOR4RegReg, node, gpr, gpr, cg);
+
+   generateLabelInstruction(LABEL, node, remStartLoop, cg);
+   generateRegRegInstruction(CMP4RegReg, node, srcOffset, length, cg);
+   generateLabelInstruction(JGE4, node, remEndLoop, cg);
+
+   generateRegMemInstruction(L1RegMem, node, gpr, generateX86MemoryReference(srcBuffer, srcOffset, 0, headerOffset, cg), cg);
+   generateMemRegInstruction(S2MemReg, node, generateX86MemoryReference(destBuffer, destOffset, 0, headerOffset, cg), gpr, cg);
+
+   // increase srcOffset by 1, destOffset by 2 to account for inflation
+   generateRegInstruction(INC4Reg, node, srcOffset, cg);
+   generateRegImmInstruction(ADD4RegImm4, node, destOffset, 2, cg);
+
+   generateLabelInstruction(JMP4, node, remStartLoop, cg);
+   generateLabelInstruction(LABEL, node, remEndLoop, cg);
```

Comment:
You should benchmark this. I think an unrolled [Duff's device](https://en.wikipedia.org/wiki/Duff%27s_device) single branch implementation may be much faster than processing one byte at a time. We implemented this on Z and had a significant performance improvement for short lengths. You can touch base with @dchopra001 to see how it is implemented.

Reply:
I believe I understand what you are suggesting, might it be more efficient to do the approach suggested by @0xdaryl, copying 8, then 4, then 2, then 1?

Reply:
Yes, it _should_ be better, but we'll have to benchmark it to find out for sure. On Z we split the code into 3 parts:

1. Process length >= 16
2. Process length >= 8 && <= 15
3. Process length < 8

We do the first two with vectors, and the last one using an unrolled Duff's device with a single branch. The entire evaluator I believe has only three branches.

Here is the benchmark we used:
https://github.com/fjeremic/jmh-benchmarks/blob/master/src/main/java/org/openj9/StringLatin1Benchmark.java

See the README in that repo on how to compile and run. It's fairly straightforward. We found on DayTrader7 lengths 0-4 account for ~80% of Strings that pass through this API, so accelerating short lengths to be as fast as possible will give us the biggest bang for the buck.

@dchopra001 could you work with @BradleyWood and sync up on all of this. Let's make sure we don't duplicate any work/investigation and help each other out.